src/Interpreters/Streaming/ChangelogQueryVisitor.cpp

```diff
@@ -131,10 +131,10 @@ void ChangelogQueryVisitorMatcher::addDeltaColumn(ASTSelectQuery & select_query,
     if (!found_delta_col)
     {
         if (is_subquery)
-            /// Need add delta if _tp_delta is not present and the @p select_query is a subquery 
+            /// Need add delta if _tp_delta is not present and the \param select_query is a subquery (not top level select)
             select_expression_list->children.emplace_back(std::make_shared<ASTIdentifier>(ProtonConsts::RESERVED_DELTA_FLAG));
-        else if (query_info.force_emit_changelog)
-            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "The query with emit changelog explicitly requires a `_tp_delta` in select list");
+        else
+            throw Exception(ErrorCodes::INCORRECT_QUERY, "The query with changelog output requires selecting the `_tp_delta` column explicitly");
```

Comment:
Look like now when top select missing `_tp_delta`, we will report this exception as well. For example, `select l.instrumentId from interesting_numbers_base_data as l join trade_summary as t on l.instrumentId = t.instrumentId` failed by default and `CREATE VIEW interesting_numbers_base_data` failed as well.

Ideally `_tp_delta` will be added back by internal logic if it is required and users didn't explicitly select it

Reply:
Yes, now explicitly select the "_tp_delta" column in the above scenario
Next stage: the behavior will be adjusted, for top selection it will always be "append only" semantics by default unless specified with "EMIT CHANGELOG"